### PR TITLE
implement deleteChunk/deleteFile for TKV engine

### PR DESCRIPTION
**Pull request by xuhui-lu**

### CODE REVIEW

File: pkg/meta/tkv.go [hunk 1]
```diff
@@ -1674,13 +1672,70 @@ func (m *kvMeta) CopyFileRange(ctx Context, fin Ino, offIn uint64, fout Ino, off
...
+			for _, val := range vals {
+				lens = append(lens, len(val)/sliceBytes)
+			}
+			return nil
+		})
+		for i, len := range lens {
+			if len == 0 {
+				continue
+			}
+			idx := binary.BigEndian.Uint32(keys[i][10:])
```
Comment (davies) [inline pkg/meta/tkv.go:1730]: 10 -> len(m.prefix)+10

File: pkg/meta/tkv.go [hunk 2]
```diff
@@ -1674,13 +1672,70 @@ func (m *kvMeta) CopyFileRange(ctx Context, fin Ino, offIn uint64, fout Ino, off
...
+	if err != syscall.Errno(0) {
+		return fmt.Errorf("delete slice from chunk %s fail: %s, retry later", key, err)
+	}
+	return nil
+}
 
-func (m *kvMeta) deleteFile(inode Ino, length uint64) {}
+func (m *kvMeta) deleteFile(inode Ino, length uint64) {
+	var indx uint32
+	for uint64(indx)*ChunkSize < length {
```
Comment (davies) [inline pkg/meta/tkv.go:1711]: tx.scanRange() to find all they chunks

File: pkg/meta/tkv.go [hunk 3]
```diff
@@ -1674,13 +1672,70 @@ func (m *kvMeta) CopyFileRange(ctx Context, fin Ino, offIn uint64, fout Ino, off
...
+	key := m.chunkKey(inode, indx)
+	var slices []*slice
+	err := m.txn(func(tx kvTxn) error {
+		buf := tx.get(key)
+		slices = readSliceBuf(buf)
+		tx.set(key, []byte{})
+		for _, s := range slices {
+			r := tx.incrBy(m.sliceKey(s.chunkid, s.size), -1)
+			if r < 0 {
+				m.deleteSlice(s.chunkid, s.size)
```
Comment (davies) [inline pkg/meta/tkv.go:1698]: This may block, so we could like to start a goroutine

File: pkg/meta/tkv.go [hunk 4]
```diff
@@ -1910,11 +1910,29 @@ func (m *kvMeta) CopyFileRange(ctx Context, fin Ino, offIn uint64, fout Ino, off
...
-func (m *dbMeta) deleteChunk(inode Ino, indx uint32) error {}
 */
 
+func (m *kvMeta) deleteChunk(inode Ino, indx uint32) error {
+	key := m.chunkKey(inode, indx)
+	var slices []*slice
+	err := m.txn(func(tx kvTxn) error {
+		buf := tx.get(key)
+		slices = readSliceBuf(buf)
+		tx.set(key, []byte{})
```
Comment (davies) [inline pkg/meta/tkv.go:1921]: tx.dels(key)

File: pkg/meta/tkv.go [hunk 5]
```diff
@@ -1940,7 +1958,33 @@ func (m *kvMeta) deleteSlice(chunkid uint64, size uint32) {
 	}
 }
 
-func (m *kvMeta) deleteFile(inode Ino, length uint64) {}
+func (m *kvMeta) deleteFile(inode Ino, length uint64) {
+	keys, err := m.scanKeys(m.chunkKey(inode, 0))
```
Comment (davies) [inline pkg/meta/tkv.go:1962]: the prefix should be  `m.fmtKey("A", inode, "C")`

File: pkg/meta/tkv.go (hunk 4 above)
Comment (xuhui-lu) [reply pkg/meta/tkv.go:1921]: hmm, to be aligned with redis implementation, should we leave the key with empty value there? Though i think there is no big difference.

File: pkg/meta/tkv.go [hunk 6]
```diff
@@ -1940,7 +1958,33 @@ func (m *kvMeta) deleteSlice(chunkid uint64, size uint32) {
...
+
+	var lens []int
+	m.txn(func(tx kvTxn) error {
+		vals := tx.gets(keys...)
+		for _, val := range vals {
+			lens = append(lens, len(val)/sliceBytes)
+		}
+		return nil
+	})
+	for i, length := range lens {
```
Comment (davies) [inline pkg/meta/tkv.go:1976]: All the returned keys should be a valid chunk, we should iterate on that.

File: pkg/meta/tkv.go (hunk 4 above)
Comment (davies) [reply pkg/meta/tkv.go:1921]: they will be leaked in TiKV. Redis will cleanup the empty keys, but other KV will not.

File: pkg/meta/tkv.go [hunk 7]
```diff
@@ -1940,7 +1958,26 @@ func (m *kvMeta) deleteSlice(chunkid uint64, size uint32) {
...
 }
 
-func (m *kvMeta) deleteFile(inode Ino, length uint64) {}
+func (m *kvMeta) deleteFile(inode Ino, length uint64) {
+	keys, err := m.scanKeys(m.fmtKey("A", inode, "C"))
+	if err != nil {
+		logger.Warnf("delete chunks of inode %d: %s", inode, err)
+		return
+	}
+	var vals [][]byte
```
Comment (davies) [inline pkg/meta/tkv.go:1967]: the `vals` is not used

File: pkg/meta/tkv.go [hunk 8]
```diff
@@ -1940,7 +1958,26 @@ func (m *kvMeta) deleteSlice(chunkid uint64, size uint32) {
...
+	m.txn(func(tx kvTxn) error {
+		vals = tx.gets(keys...)
+		return nil
+	})
+	for i := range vals {
```
Comment (davies) [inline pkg/meta/tkv.go:1972]: use `keys` should be the same

**Review by davies — APPROVED** [review]

**Pull request merged**